malariagen_data/af1.py

```diff
@@ -0,0 +1,747 @@
+            metadata e.g., "country == 'Burkina Faso'".
+        site_mask : {"funestus"}
+            Site filters mask to apply.
+        min_minor_ac : int, optional
+            The minimum minor allele count. SNPs with a minor allele count
+            below this value will be excluded prior to thinning.
+        max_missing_an : int, optional
+            The maximum number of missing allele calls to accept. SNPs with
+            more than this value will be excluded prior to thinning. Set to 0
+            (default) to require no missing calls.
+        n_components : int, optional
+            Number of components to return.
+
+        Returns
+        -------
+        df_pca : pandas.DataFrame
+            A dataframe of sample metadata, with columns "PC1", "PC2", "PC3",
+            etc., added.
+        evr : np.ndarray
+            An array of explained variance ratios, one per component.
+
+        Notes
+        -----
+        This computation may take some time to run, depending on your computing
+        environment. Results of this computation will be cached and re-used if
+        the `results_cache` parameter was set when instantiating the Ag3 class.
+
+        """
+        debug = self._log.debug
+
+        # change this name if you ever change the behaviour of this function, to
+        # invalidate any previously cached data
+        name = PCA_RESULTS_CACHE_NAME
+
+        debug("normalize params for consistent hash value")
+        params = dict(
+            region=self.resolve_region(region).to_dict(),
+            n_snps=n_snps,
+            thin_offset=thin_offset,
+            sample_sets=self._prep_sample_sets_arg(sample_sets=sample_sets),
+            sample_query=sample_query,
+            site_mask=site_mask,
+            min_minor_ac=min_minor_ac,
+            max_missing_an=max_missing_an,
+            n_components=n_components,
+        )
+
+        debug("try to retrieve results from the cache")
+        try:
+            results = self.results_cache_get(name=name, params=params)
+
+        except CacheMiss:
+            results = self._pca(**params)
+            self.results_cache_set(name=name, params=params, results=results)
+
+        debug("unpack results")
+        coords = results["coords"]
+        evr = results["evr"]
+
+        debug("add coords to sample metadata dataframe")
+        df_samples = self.sample_metadata(
+            sample_sets=sample_sets,
+            sample_query=sample_query,
+        )
+        df_coords = pd.DataFrame(
+            {f"PC{i + 1}": coords[:, i] for i in range(n_components)}
+        )
+        df_pca = pd.concat([df_samples, df_coords], axis="columns")
+
+        return df_pca, evr
+
+    def plot_snps_track(
+        self,
+        region,
+        sample_sets=None,
+        sample_query=None,
+        site_mask=DEFAULT_SITE_MASK,
+        width=800,
+        height=120,
+        max_snps=200_000,
+        show=True,
+    ):
```

Comment:
There are several functions like this, where the implementation is fully generic and could be moved to the base class, but there is a parameter (in thise case `site_mask`) where the default value needs to be different in the child classes.

It would be good to move these to the base class somehow. There's several ways I could think of doing it. Perhaps the simplest would be to set the default value of the `site_mask` parameter to the string "default". Then the implementation could check if the value is "default" and replace it with the value of a class or instance attribute (e.g., `self._default_site_mask`) which is set in the child class.

Reply:
Yes, I agree, only I feel it might be slightly cleaner to default to `None` in the parent class, then check for `None` and replace that with the value an abstract property set by the child class, as you suggest. Thanks!

Reply:
In the case of `plot_snps_track`, the subclasses actually have non-generic `cols` and `tooltips` too, but perhaps it will be OK to solve those in a similar way. (I wanted to avoid refactoring any functions that required non-trivial work to generalise, to avoid opening cans of worms and delaying the Af1 release, but now I appreciate that these cases are not very far off. Will do.)

Reply:
> Yes, I agree, only I feel it might be slightly cleaner to default to `None` in the parent class, then check for `None` and replace that with the value an abstract property set by the child class, as you suggest. Thanks!

Not sure exactly what you mean here, but just flagging that we still need to allow for user to supply None meaning apply no site mask at all.

Reply:
That's a good point, thanks. (Sorry, I didn't mean to prompt any responses this week, so extra thanks!)

Reply:
FWIW, I'll use "default" as the default value instead of `None`, as you originally suggested, and I've checked that specifying `site_mask=None` when calling does indeed override the default of "default" to be `None`, rather than defaulting to "default" instead of `None`, if that makes sense.